canonical-web-and-design/maas-ui
# feat(ui): make machine action menu reusable across all node types

## Done

- Refactored the machine `TakeActionMenu` so it could be used across node types, and moved it from `/machines/components/TakeActionMenu` to `/base/components/NodeActionMenu`
- Added some basic node utils for identifying its type

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to /MAAS/r/devices and check that hovering over the action menu shows a tooltip
- "Select" some devices by opening `ui/src/app/devices/views/DeviceList/DeviceListHeader/DeviceListHeader.tsx` and editing line 41 to:
  ``` ts
  nodes={devices}
  ```
- Open the action menu and check that the action counts are correct
- Check that clicking the action names updates the header (the actual forms themselves will be implemented later)
- Go to the machine list and check that opening action forms works as before
- Go to a machine's details page and check that opening action forms works as before
- Got a LXD KVM host's details page and check that opening action forms works as before

## Fixes

Fixes canonical-web-and-design/app-tribe#524

## Launchpad issue

[lp#1951893](https://bugs.launchpad.net/maas/+bug/1951893)

## Screenshot
![Screenshot 2021-11-23 at 10-52-38 Devices bolla MAAS](https://user-images.githubusercontent.com/25733845/142957726-fde77ab4-d3f4-4845-9aa8-bb0b4ce1ea52.png)
